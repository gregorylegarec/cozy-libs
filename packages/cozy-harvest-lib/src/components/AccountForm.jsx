import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import Manifest from '../Manifest'

const legacyOptions = options =>
  options.map(option => ({ ...option, label: option.name }))

export class AccountField extends PureComponent {
  render() {
    const { name, t, type } = this.props
    const label = t(`fields.${name}.label`)
    const fieldProps = {
      ...this.props,
      className: 'u-m-0', // 0 margin
      fullwidth: true,
      label,
      size: 'medium'
    }
    switch (type) {
      case 'dropdown':
        return (
          <Field
            {...fieldProps}
            type="dropdown"
            options={legacyOptions(fieldProps.options)}
          />
        )
      case 'password':
        return <Field {...fieldProps} />
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

export class AccountFields extends PureComponent {
  render() {
    const { manifestFields, t } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField key={index} name={field.name}>
            {({ input }) => (
              <AccountField label={field.name} {...field} {...input} t={t} />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { lang, locales } = props
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  render() {
    const { account, fields, t } = this.props
    const sanitizedFields = Manifest.sanitizeFields(fields)
    const initialValues = account ? account.auth : {}
    return (
      <Form
        initialValues={initialValues}
        // eslint-disable-next-line no-console
        onSubmit={v => console.log(v)}
        render={({ values }) => (
          <div>
            <AccountFields manifestFields={sanitizedFields} t={t} />
            <Button
              className="u-mt-2 u-mb-1-half"
              extension="full"
              label={t('accountForm.submit.label')}
              onclick={() => alert(JSON.stringify(values, 0, 2))}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  fields: PropTypes.object.isRequired,
  locales: PropTypes.object
}

export default translate()(AccountForm)
