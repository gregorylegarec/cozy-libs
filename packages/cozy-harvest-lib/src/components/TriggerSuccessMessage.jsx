import React, { Component } from 'react'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate } from 'cozy-ui/react/I18n'

import connectingImg from '../assets/images/connecting.svg'

const IDLE = 'idle'
const DONE = 'done'

export class TriggerSuccessMessage extends Component {
  state = {
    status: IDLE
  }

  onClick = () => {
    const { status } = this.state
    if (status !== DONE) {
      const { onDone } = this.props
      onDone()
    }
    this.setState({ status: DONE })
  }

  render() {
    const { t } = this.props
    const { status } = this.state
    const isDone = status === DONE
    return (
      <div className="u-ta-center">
        <img src={connectingImg} />
        <h2>{t('triggerSuccessMessage.title')}</h2>
        <p className="u-ta-left">{t('triggerSuccessMessage.description')}</p>
        <Button
          disabled={isDone}
          label={t('triggerSuccessMessage.button.label')}
          onClick={this.onClick}
        />
      </div>
    )
  }
}

TriggerSuccessMessage.propTypes = {
  onDone: PropTypes.func.isRequired
}

export default translate()(TriggerSuccessMessage)
