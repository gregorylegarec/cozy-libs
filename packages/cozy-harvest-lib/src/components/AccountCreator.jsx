import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { prepareAccountData } from '../helpers/accounts'

/**
 * Encapsulates an AccountForm and create an account with resulting data.
 */
export class AccountCreator extends PureComponent {
  create = async data => {
    const { createAccount, konnector, onCreate, onCreateSuccess } = this.props
    onCreate()
    const account = await createAccount(prepareAccountData(konnector, data))
    onCreateSuccess(account)
  }

  render() {
    const { konnector, submitting } = this.props
    return (
      <AccountForm
        fields={konnector.fields}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.create}
        submitting={submitting}
      />
    )
  }
}

AccountCreator.propTypes = {
  createAccount: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  submitting: PropTypes.bool,
  onCreate: PropTypes.func.isRequired,
  onCreateSuccess: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountCreator)
