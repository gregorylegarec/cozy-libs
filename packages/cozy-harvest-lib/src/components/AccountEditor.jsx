import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { hydrateAccount } from '../helpers/accounts'

/**
 * Encapsulates an AccocuntForm of an existing account, and allow to update it.
 */
export class AccountEditor extends PureComponent {
  update = async data => {
    const { account, onUpdate, onUpdateSuccess, updateAccount } = this.props
    onUpdate()
    const updatedAccount = await updateAccount(hydrateAccount(account, data))
    console.debug({ updatedAccount })
    onUpdateSuccess(updatedAccount)
  }

  render() {
    const { account, konnector, submitting } = this.props
    console.debug('AcconutEditor.render', { account })
    return (
      <AccountForm
        fields={konnector.fields}
        initialValues={account && (account.auth || account.oauth)}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.update}
        submitting={submitting}
      />
    )
  }
}

AccountEditor.propTypes = {
  account: PropTypes.object,
  createAccount: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  submitting: PropTypes.bool,
  onUpdate: PropTypes.func.isRequired,
  onUpdateSuccess: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountEditor)
