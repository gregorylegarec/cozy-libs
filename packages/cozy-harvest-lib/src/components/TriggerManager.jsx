import React, { Component } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountCreator from './AccountCreator'
import AccountEditor from './AccountEditor'
import TriggerSuccessMessage from './TriggerSuccessMessage'
import { triggersMutations } from '../connections/triggers'
import {
  buildKonnectorCron,
  buildKonnectorTriggerAttributes
} from '../helpers/triggers'

const IDLE = 'IDLE'
const RUNNING = 'RUNNING'
const LOGGED = 'LOGGED'
const SUCCESS = 'SUCCESS'

/**
 * Deals with konnector configuration, i.e encapsulate account creation or
 * edition
 * @type {Component}
 */
export class TriggerManager extends Component {
  state = {
    status: IDLE
  }

  /**
   * Creation/update start handler
   * Set the status to RUNNING as soon as an account is being created
   * or updated.
   */
  handleAccountMutation = () =>
    this.setState({
      status: RUNNING
    })

  /**
   * Account creation success handler
   * @param  {Object}  account Created io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  handleTriggerSuccessMessage = async account => {
    const { createTrigger, konnector } = this.props

    const trigger = await createTrigger(
      buildKonnectorTriggerAttributes({
        konnector,
        account,
        cron: buildKonnectorCron(konnector)
      })
    )

    return await this.launch(trigger)
  }

  /**
   * Account update success handler
   * @param  {Object}  account Updated io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  handleAccountUpdateSuccess = async account => {
    this.setState({
      updatedAccount: account
    })

    const { trigger } = this.props
    return await this.launch(trigger)
  }

  /**
   * Launches a trigger
   * @param  {Object}  trigger io.cozy.triggers document
   * @return {Promise}         [description]
   */
  launch = async trigger => {
    const { launchTrigger, onLoginSuccess, waitForLoginSuccess } = this.props

    const job = await waitForLoginSuccess(await launchTrigger(trigger))

    if (['queued', 'running'].includes(job.state)) {
      onLoginSuccess(trigger)
    }

    this.setState({
      status: SUCCESS
    })
  }

  render() {
    const { account, konnector, onDone } = this.props
    const { status, updatedAccount } = this.state
    const succeed = [LOGGED, SUCCESS].includes(status)
    const submitting = status === RUNNING

    return account ? (
      <AccountEditor
        account={updatedAccount || account}
        konnector={konnector}
        onUpdate={this.handleAccountMutation}
        onUpdateSuccess={this.handleAccountUpdateSuccess}
        submitting={submitting}
      />
    ) : succeed ? (
      <TriggerSuccessMessage onDone={onDone} />
    ) : (
      <AccountCreator
        konnector={konnector}
        onCreate={this.handleAccountMutation}
        onCreateSuccess={this.handleTriggerSuccessMessage}
        submitting={submitting}
      />
    )
  }
}

TriggerManager.propTypes = {
  account: PropTypes.object,
  createTrigger: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  launchTrigger: PropTypes.func.isRequired,
  onDone: PropTypes.func.isRequired,
  onLoginSuccess: PropTypes.func.isRequired,
  trigger: PropTypes.object,
  waitForLoginSuccess: PropTypes.func.isRequired
}

export default withMutations(triggersMutations)(TriggerManager)
