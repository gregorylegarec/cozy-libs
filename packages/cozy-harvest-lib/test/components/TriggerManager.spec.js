/* eslint-env jest */
import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { TriggerManager } from 'components/TriggerManager'

configure({ adapter: new Adapter() })

jest.mock('cozy-client', () => ({
  withMutations: () => jest.fn()
}))

const triggersHelper = require('helpers/triggers')
triggersHelper.buildKonnectorCron = jest.fn().mockReturnValue('0 0 0 * * 0')

const fixtures = {
  data: {
    username: 'foo',
    passphrase: 'bar'
  },
  konnector: {
    slug: 'konnectest'
  },
  triggerAttributes: {
    arguments: '0 0 0 * * 0',
    type: '@cron',
    worker: 'konnector',
    worker_arguments: {
      account: 'a87f9a8bd3884479a48811e7b7deec75',
      konnector: 'konnectest'
    }
  },
  createdAccount: {
    _id: 'a87f9a8bd3884479a48811e7b7deec75',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    }
  },
  createdTrigger: {
    id: '669e9a7cc3064a97bc0aa20feef71cb2',
    _type: 'io.cozy.triggers',
    attributes: {
      arguments: '0 0 0 * * 0',
      type: '@cron',
      worker: 'konnector',
      worker_arguments: {
        account: 'a87f9a8bd3884479a48811e7b7deec75',
        konnector: 'konnectest'
      }
    }
  },
  launchedJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    attributes: {
      domain: 'cozy.tools:8080',
      worker: 'konnector',
      state: 'running',
      queued_at: '2016-09-19T12:35:08Z',
      started_at: '2016-09-19T12:35:08Z',
      error: ''
    },
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  },
  runningJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    attributes: {
      domain: 'cozy.tools:8080',
      worker: 'konnector',
      state: 'running',
      queued_at: '2016-09-19T12:35:08Z',
      started_at: '2016-09-19T12:35:08Z',
      error: ''
    },
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  }
}

const createAccountMock = jest.fn().mockResolvedValue(fixtures.createdAccount)
const createTriggerMock = jest.fn().mockResolvedValue(fixtures.createdTrigger)
const launchTriggerMock = jest.fn().mockResolvedValue(fixtures.launchedJob)
const waitForLoginSuccessMock = jest.fn().mockResolvedValue(fixtures.runningJob)

const onDoneMock = jest.fn()

describe('TriggerManager', () => {
  beforeEach(() => {
    createAccountMock.mockClear()
    createTriggerMock.mockClear()
    launchTriggerMock.mockClear()
    onDoneMock.mockClear()
    waitForLoginSuccessMock.mockClear()
  })

  it('should render', () => {
    const component = shallow(
      <TriggerManager
        createAccount={createAccountMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should pass account auth values', () => {
    const account = {
      auth: {
        username: 'foo',
        passphrase: 'bar'
      }
    }
    const component = shallow(
      <TriggerManager
        account={account}
        createAccount={createAccountMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should pass oauth values', () => {
    const account = {
      oauth: {
        accessToken: 'abcdef12345'
      }
    }

    const component = shallow(
      <TriggerManager
        account={account}
        createAccount={createAccountMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should call createAccount on submit', () => {
    const wrapper = shallow(
      <TriggerManager
        createAccount={createAccountMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
      />
    )

    wrapper.instance().onSubmit(fixtures.data)

    expect(createAccountMock).toHaveBeenCalledWith({
      account_type: 'konnectest',
      auth: {
        username: 'foo',
        passphrase: 'bar'
      }
    })
  })

  it('should call createTrigger on submit', async () => {
    const wrapper = shallow(
      <TriggerManager
        createAccount={createAccountMock}
        createTrigger={createTriggerMock}
        launchTrigger={launchTriggerMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
        waitForLoginSuccess={waitForLoginSuccessMock}
      />
    )

    await wrapper.instance().onSubmit(fixtures.data)

    expect(createTriggerMock).toHaveBeenCalledWith(fixtures.triggerAttributes)
  })

  it('should call launchTrigger on submit', async () => {
    const wrapper = shallow(
      <TriggerManager
        createAccount={createAccountMock}
        createTrigger={createTriggerMock}
        launchTrigger={launchTriggerMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
        waitForLoginSuccess={waitForLoginSuccessMock}
      />
    )

    await wrapper.instance().onSubmit(fixtures.data)

    expect(launchTriggerMock).toHaveBeenCalledWith(fixtures.createdTrigger)
  })

  it('should call waitForLoginSuccess on submit', async () => {
    const wrapper = shallow(
      <KonnectorManager
        createAccount={createAccountMock}
        createTrigger={createTriggerMock}
        launchTrigger={launchTriggerMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
        waitForLoginSuccess={waitForLoginSuccessMock}
      />
    )

    await wrapper.instance().onSubmit(fixtures.data)

    expect(waitForLoginSuccessMock).toHaveBeenCalledWith(fixtures.launchedJob)
  })

  it('should update state on submit', () => {
    const wrapper = shallow(
      <TriggerManager
        createAccount={createAccountMock}
        konnector={fixtures.konnector}
        onDone={onDoneMock}
      />
    )

    wrapper.instance().onSubmit(fixtures.data)

    expect(wrapper.state().status).toBe('CREATING')
  })

  it('should render as expected after submit success', async () => {
    const wrapper = shallow(
      <TriggerManager
        createAccount={createAccountMock}
        createTrigger={createTriggerMock}
        konnector={fixtures.konnector}
        launchTrigger={launchTriggerMock}
        onDone={onDoneMock}
        waitForLoginSuccess={waitForLoginSuccessMock}
      />
    )

    await wrapper.instance().onSubmit(fixtures.data)
    const component = wrapper.getElement()

    expect(wrapper.state().status).toBe('CREATION_SUCCESS')
    expect(component).toMatchSnapshot()
  })
})
