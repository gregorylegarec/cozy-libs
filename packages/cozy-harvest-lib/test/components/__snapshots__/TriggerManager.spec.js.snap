// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KonnectorManager should pass account auth values 1`] = `
<Wrapper
  account={
    Object {
      "auth": Object {
        "passphrase": "bar",
        "username": "foo",
      },
    }
  }
  initialValues={
    Object {
      "passphrase": "bar",
      "username": "foo",
    }
  }
  onSubmit={[Function]}
  submitting={false}
/>
`;

exports[`KonnectorManager should pass oauth values 1`] = `
<Wrapper
  account={
    Object {
      "oauth": Object {
        "accessToken": "abcdef12345",
      },
    }
  }
  initialValues={
    Object {
      "accessToken": "abcdef12345",
    }
  }
  onSubmit={[Function]}
  submitting={false}
/>
`;

exports[`KonnectorManager should render 1`] = `
<Wrapper
  onSubmit={[Function]}
  submitting={false}
/>
`;

exports[`KonnectorManager should render as expected after submit success 1`] = `
<Wrapper
  onDone={[Function]}
/>
`;
